unitary/examples/quantum_chinese_chess/chess.py

```diff
@@ -476,6 +474,56 @@ def game_over(self) -> None:
         # TODO(): add the following checks
         # - If player 0 made N repeatd back-and_forth moves in a row.
 
+    def save_snapshot(self) -> None:
+        """Saves the current length of the effect history, qubit_remapping_dict, and the current classical states of all pieces."""
+        # Save the current length of the effect history and qubit_remapping_dict.
+        self.board.board.save_snapshot()
+
+        # Save the classical states of all pieces.
+        snapshot = []
+        for row in range(10):
+            for col in "abcdefghi":
+                piece = self.board.board[f"{col}{row}"]
+                snapshot.append(
+                    [piece.type_.value, piece.color.value, piece.is_entangled]
+                )
+        self.classical_properties_history.append(snapshot)
+
+    def undo(self) -> bool:
+        """Undo the last move, which includes reset quantum effects and classical properties, and remapping
+        qubits.
+
+        Returns True if the undo is success, and False otherwise.
+        """
+        world = self.board.board
+        if (
+            len(world.effect_history_length) <= 1
+            or len(world.qubit_remapping_dict_length) <= 1
+            or len(self.classical_properties_history) <= 1
+        ):
+            # length == 1 corresponds to the initial state, and no more undo could be made.
+            return False
+
+        # Recover the mapping of qubits to the last snapshot, remove any related post selection memory,
+        # and recover the effects up to the last snapshot (which was saved after the last move finished).
+        try:
+            world.restore_last_snapshot()
+        except:
+            return False
+
+        # Recover the classical properties of all pieces to the last snapshot.
+        self.classical_properties_history.pop()
+        snapshot = self.classical_properties_history[-1]
+        index = 0
+        for row in range(10):
+            for col in "abcdefghi":
```

Comment:
you can flatten this with:

```py
for index, (row, col) in enumerate(itertools.product(range(10), "abcdefghi")):
    # do stuff
```

though this might be crossing into unreasonable python hackery territory :)

Reply:
Thanks for the suggestion. I did a test trying to compare their speeds:

In [13]: s1=time.time()
    ...: for a, b in itertools.product(range(10), "abcdefghi"):
    ...:     c,d = a,b
    ...: s2 = time.time()
    ...: 
    ...: 
    ...: e1=time.time()
    ...: for row in range(10):
    ...:     for col in "abcdefghi":
    ...:         c,d = row, col
    ...: e2 = time.time()

In [14]: s2-s1
Out[14]: 0.00016570091247558594

In [15]: e2-e1
Out[15]: 0.00012350082397460938

I tried twice and double loops both wins in this case. I'll keep it like this for now. Thanks.